backend/src/main/scala/com/softwaremill/bootzooka/user/application/UserService.scala

```diff
@@ -61,7 +63,23 @@ class UserService(
   def authenticate(login: String, nonEncryptedPassword: String): Future[Option[BasicUserData]] =
     userDao
       .findByLoginOrEmail(login)
-      .map(userOpt => userOpt.filter(u => User.passwordsMatch(nonEncryptedPassword, u)).map(BasicUserData.fromUser))
+      .map(_.filter(u => hashing.verifyPassword(u.password, nonEncryptedPassword, u.salt)))
+      .flatMap {
+        case Some(u) => rehashIfRequired(u, nonEncryptedPassword)
+        case None    => Future.successful(None)
+      }
+      .map(_.map(BasicUserData.fromUser))
+
+  def rehashIfRequired(u: User, nonEncryptedPassword: String): Future[Option[User]] =
```

Comment:
some explanation on the rehashing mechanism would be good :)

Reply:
done